include/openthread/platform/infra_if.h

```diff
@@ -100,6 +90,29 @@ extern void otPlatInfraIfRecvIcmp6Nd(otInstance *        aInstance,
                                      const uint8_t *     aBuffer,
                                      uint16_t            aBufferLength);
 
+/**
+ * The infra interface driver calls this method to notify OpenThread
+ * of the interface state changes.
+ *
+ * @param[in]  aInstance          The OpenThread instance structure.
+ * @param[in]  aInfraIfIndex      The index of the infrastructure interface.
+ * @param[in]  aIsRunning         A boolean that indicates whether the infrastructure
+ *                                interface is running.
+ * @param[in]  aLinkLocalAddress  A pointer to the IPv6 link-local address of the infrastructure
+ *                                interface. NULL if the IPv6 link-local address is lost.
+ *
+ * @retval  OT_ERROR_NONE           Successfully updated the infra interface status.
+ * @retval  OT_ERROR_INVALID_STATE  The Routing Manager is not initialized.
+ * @retval  OT_ERROR_INVALID_ARGS   The @p aInfraIfIndex doesn't match the infra interface the
+ *                                  Routing Manager are initialized with, or the @p aLinkLocalAddress
+ *                                  is not a valid IPv6 link-local address.
+ *
```

Comment:
This is a callback from the perspective of platform layer so that it can inform OT core about state change. Is there any behavior that we expect the platform layer to do in case of an error?

Can we consider not returning an `error`?

Reply:
The minor issue I can see without returning an `error` is that the user code might failed to call this method with correct network interface and link-local address, and didn't realize that. The `error` can help users to find such incorrect behavior easier.

Reply:
I share the similar thoughts with @simonlingoogle. The platform implementation may invoke this method in invalid state or with invalid arguments and because this function is public API, I think it is better to return the error but not assert or simply die.

We simply die on this error for posix implementation because we are sure that we are providing the correct arguments.
```c++
SuccessOrDie(otPlatInfraIfStateChanged(aInstance, sInfraIfIndex, platformInfraIfIsRunning(),
                                                   platformInfraIfGetLinkLocalAddress()));
```